File: QID-986-MVAnpcatime/MVAnpcatime.r

# clear all variables
rm(list = ls(all = TRUE))
graphics.off()

# load data
x  = read.table("timebudget.dat")
n  = nrow(x)                            # number of rows
p  = ncol(x)                            # number of columns
x1 = sqrt((n - 1) * apply(x, 2, var))   # estimated std error for each variable
x2 = x - matrix(apply(as.matrix(x), 2, mean), nrow = n, ncol = p, byrow = T)  # X - mean(X)
x  = as.matrix(x2/matrix(x1, nrow = n, ncol = p, byrow = T))                  # Standardizes Data
e  = eigen(x %*% t(x)/n)                # Eigenvalues/Eigenvectors
e1 = e$values
e2 = e$vectors
a  = e2[, 1:2]                          # first two Eigenvectors corresponding to largest 2 Eigenvalues
w  = a * sqrt(matrix(e1[1:2], nrow(a), ncol(a), byrow = TRUE))                # coordinates of individuals

# Plot 1: the representation of the individuals
plot(w, type = "n", xlab = "First Factor - Individuals", ylab = "Second Factor - Individuals", 
    main = "Time Budget data", cex.lab = 1.2, cex.axis = 1.2, cex.main = 1.8, lwd = 2)
label = c("maus", "waus", "wnus", "mmus", "wmus", "msus", "wsus", "mawe", "wawe", 
    "wnwe", "mmwe", "wmwe", "mswe", "wswe", "mayo", "wayo", "wnyo", "mmyo", "wmyo", 
    "msyo", "wsyo", "maes", "waes", "wnes", "mmes", "wmes", "mses", "wses")
text(w, label, cex = 1.2)
abline(h = 0, v = 0)

g  = eigen(t(x) %*% x)  # Eigenvalues/Eigenvectors
g1 = g$values
g2 = g$vectors
b  = g2[, 1:2]          # first 2 vectors corresponding to largest 2 Eigenvalues
z  = b * sqrt(matrix(g1[1:2], nrow(b), ncol(b), byrow = TRUE))  # coordinates of variables
dev.new()

# Plot 2: the representation of the variables
ucircle = cbind(cos((0:360)/180 * pi), sin((0:360)/180 * pi))
plot(ucircle, type = "l", lty = "solid", col = "blue", xlab = "First Factor - Expenditures", 
    ylab = "Second Factor - Expenditures", main = "Time Budget data", cex.lab = 1.2, 
    cex.axis = 1.2, cex.main = 1.8, lwd = 2)
abline(h = 0, v = 0)
label = c("prof", "tran", "hous", "kids", "shop", "pers", "eati", "slee", "tele", 
    "leis")
text(z, label, cex = 1.2) 

File: QID-986-MVAnpcatime/MVAnpcatime.py
#works on numpy 1.25.2, pandas 2.1.1 and matplotlib 3.8.0
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Circle

x = pd.read_csv('timebudget.dat', sep="\s+", header = None)

n, p = x.shape
x = (x-x.mean())/(np.sqrt((n-1) * np.var(x, axis=0)))
e, v = np.linalg.eig(np.dot(x, x.T/n))
w = v[:,0:2] * np.sqrt(e[0:2])

labels = ["maus", "waus", "wnus", "mmus", "wmus", "msus", "wsus", "mawe", "wawe", 
    "wnwe", "mmwe", "wmwe", "mswe", "wswe", "mayo", "wayo", "wnyo", "mmyo", "wmyo", 
    "msyo", "wsyo", "maes", "waes", "wnes", "mmes", "wmes", "mses", "wses"]

fig, ax = plt.subplots(figsize=(8, 6))
ax.plot([0,0],[-0.6,0.6], color='grey', alpha=0.5)
ax.plot([-1,0.6],[0,0], color='grey', alpha=0.5)
ax.set_xlim(-0.17, 0.12)
ax.set_ylim(-0.11, 0.1)
ax.scatter(w[:,0], w[:,1])
[ax.annotate(name, (w[i, 0], w[i, 1])) for i, name in enumerate(labels)]
ax.set_title('Time Budget Data')
ax.set_xlabel('First Factor - Individuals')
ax.set_ylabel('Second Factor - Individuals')
plt.show()

e1, v1 = np.linalg.eig(np.dot(x.T, x))
z = v1[:,0:2] * np.sqrt(e1[0:2])

labels = ["prof", "tran", "hous", "kids", "shop", "pers", "eati", "slee", "tele", 
    "leis"]

fig, ax = plt.subplots(figsize=(8, 8))
ax.plot([0,0],[-1,1], color='grey', alpha=0.5)
ax.plot([-1,1],[0,0], color='grey', alpha=0.5)
ax.set_ylim(-1,1)
ax.set_xlim(-1,1)
circle = Circle((0,0), 1, color='b', fill=False)
ax.add_patch(circle)
ax.set_xticks([-1,-0.5,0,0.5,1])
ax.set_yticks([-1,-0.5,0,0.5,1])
ax.set_aspect('equal')
ax.scatter(z[:,0], z[:,1])
[ax.annotate(name, (z[i, 0], z[i, 1])) for i, name in enumerate(labels)]
ax.set_title('Time Budget Data')
ax.set_xlabel('First Factor - Expenditures')
ax.set_ylabel('Second Factor - Expenditures')
plt.show()
